asio/include/asio/experimental/impl/promise.hpp

```diff
@@ -64,10 +64,10 @@ struct promise_impl<void(Ts...), Executor, Allocator>
   Executor executor;
 
   template<typename Func, std::size_t... Idx>
-  void apply_impl(Func f, asio::detail::index_sequence<Idx...>)
+  void apply_impl(Func&& f, asio::detail::index_sequence<Idx...>)
   {
     auto& result_type = *reinterpret_cast<promise_impl::result_type*>(&result);
-    f(std::get<Idx>(std::move(result_type))...);
+    std::forward<Func>(f)(std::get<Idx>(std::move(result_type))...);
```

Comment:
This part makes no sense

Reply:
Either that or take by value (i.e. move-construct the function) and move. Everything else does not work with rvalue-ref-qualified call operators on completion handlers. Or am I missing something?

If it's between these two options, I'd avoid the (unnecessary) move-construction.